# Читання даних з веб-сторінок

---

## 0. Інсталяція потрібних пакетів

Для початку варто інсталювати пакети `rvest`, `dplyr`, `stringr`.

In [1]:
# install.packages("rvest") # парсер вбесторінок
# install.packages("dplyr") # маніпулювання та вибірка даних
# install.packages("stringr") # робота з текстом/рядками

Завантажуємо пакет у память, щоб можна було використовувати функції.

In [2]:
library(rvest)
library(dplyr)
library(stringr)


Attaching package: 'dplyr'


The following objects are masked from 'package:stats':

    filter, lag


The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union




## 1. Читаємо вміст вебсторінки та потрібні дані

Вказуємо адресу ресурсу для читання даних:

In [3]:
url <- "https://abit-poisk.org.ua/rate2022/direction/968614"
#url <- "https://abit-poisk.org.ua/rate2022/direction/984185"
url

[1] "https://abit-poisk.org.ua/rate2022/direction/984185"

Читаємо сторінку у змінну у форматі `html` як набір тексту:

In [4]:
page <- read_html(url)
page

{html_document}
<html lang="uk">
[1] <head>\n<meta http-equiv="Content-Type" content="text/html; charset=UTF-8 ...
[2] <body>\n<div id="js-ba-options" data-options='{"banners":[["\\/assets\\/i ...

Витягуємо з інформації про сторінку усі потрібні нам теги, у даному випадку це таблиці, тег `table`.

In [5]:
tables <- html_nodes(page, "table")
tables

{xml_nodeset (2)}
[1] <table class="table table-bordered table-hover">\n<thead><tr>\n<th class= ...
[2] <table class="table table-bordered table-hover" style="padding-right: 0;p ...

Конвертуємо отриману таблицю з формату html у `data.frame` та переглядаємо перших кілька рядків таблиці, щоб переконатися, що усе `ОК`.

In [6]:
data <- as.data.frame(html_table(tables[1]), fill = T)
data

X. 
1   1  
2   2  
3   3  
4   4  
5   5  
6   6  
7   7  
8   8  
9   9  
10  10 
11  11 
12  12 
13  13 
14  14 
15  15 
16  16 
17  17 
18  18 
19  19 
20  20 
21  21 
22  22 
23  23 
24  24 
25  25 
26  26 
27  27 
28  28 
29  29 
30  30 
... ...
162 161
163 162
164 163
165 164
166 165
167 166
168 167
169 168
170 169
171 170
172 171
173 172
174 173
175 174
176 175
177 176
178 177
179 178
180 179
181 180
182 181
183 182
184 183
185 184
186 185
187 186
188 187
189 188
190 189
191 190
    ПІБ                                                                                                                                                                                                                          
1   1\n                                                        \n                                120-7964783  В. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1   
2   2\n                                                        \n                                Горецька Н. Л.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
3   3\n                                                        \n                                Жарчинський М. С.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1    
4   4\n                                                        \n                                Пильгун А. Ю.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1        
5   5\n                                                        \n                                Ошейко В. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 2         
6   6\n                                                        \n                                Круш І. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 4           
7   7\n                                                        \n                                Строзюк Р. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1        
8   8\n                                                        \n                                Лавренюк А. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
9   9\n                                                        \n                                Дігалевич І. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1      
10  10\n                                                        \n                                Сівачук М. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
11  11\n                                                        \n                                Свирида А. Ю.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
12  12\n                                                        \n                                Мельник Д. А.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 2       
13  13\n                                                        \n                                Бондар А. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 5        
14  14\n                                                        \n                                120-7996285  Б. М.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1  
15  15\n                          

## 2. Очистка та трансформація даних

### 2.1. Видалення зайвих стовпців

Видалимо 3-й та 7-й стовпець з інформацією про квоту

In [7]:
data |> select(c(3,7)) |> head()

П Квота  
1 1 Квота-2
2 1 Квота-1
3 1 —      
4 1 —      
5 2 —      
6 4 —

In [8]:
data <- data |> select(-c(3,7))
data |> head()

X.
1 1 
2 2 
3 3 
4 4 
5 5 
6 6 
  ПІБ                                                                                                                                                                                                                       
1 1\n                                                        \n                                120-7964783  В. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1
2 2\n                                                        \n                                Горецька Н. Л.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1    
3 3\n                                                        \n                                Жарчинський М. С.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1 
4 4\n                                                        \n                                Пильгун А. Ю.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1     
5 5\n                                                        \n                                Ошейко В. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 2      
6 6\n                                                        \n                                Круш І. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 4        
  Заг.бал Статус       
1 176.100 До наказу (б)
2 131.900 До наказу (б)
3 195.200 До наказу (б)
4 192.400 До наказу (б)
5 188.100 До наказу (б)
6 185.100 До наказу (б)
  Складові.заг..балу                                                                                                                                                                                                                                                                                                                                          
1 187  176  160                                \n                                \n                                    Українська мова: 187\n\n                                                                                    Математика: 176\n\n                                                                                    Історія України: 160
2 138  125  140                                \n                                \n                                    Українська мова: 138\n\n                                                                                    Математика: 125\n\n                                                                                    Історія України: 140
3 184  200  200                                \n                                \n                                    Українська мова: 184\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 200
4 184  200  186                                \n                                \n                                    Українська мова: 184\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 186
5 187  200  160                                \n                                \n                                    Українська мова: 187\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 160
6 177  192  180                                \n                                \n                           

### 2.2. Іменування стовпців

Замінимо назви стовпців дата-фрейму для зручності використання під час роботи з даними

In [9]:
colnames(data) <- c("Rank", "Name", "TotalGrade", "StudyType", "Grades", "IsStudent")
data |> head()

Rank
1 1   
2 2   
3 3   
4 4   
5 5   
6 6   
  Name                                                                                                                                                                                                                      
1 1\n                                                        \n                                120-7964783  В. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1
2 2\n                                                        \n                                Горецька Н. Л.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1    
3 3\n                                                        \n                                Жарчинський М. С.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1 
4 4\n                                                        \n                                Пильгун А. Ю.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1     
5 5\n                                                        \n                                Ошейко В. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 2      
6 6\n                                                        \n                                Круш І. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 4        
  TotalGrade StudyType    
1 176.100    До наказу (б)
2 131.900    До наказу (б)
3 195.200    До наказу (б)
4 192.400    До наказу (б)
5 188.100    До наказу (б)
6 185.100    До наказу (б)
  Grades                                                                                                                                                                                                                                                                                                                                                      
1 187  176  160                                \n                                \n                                    Українська мова: 187\n\n                                                                                    Математика: 176\n\n                                                                                    Історія України: 160
2 138  125  140                                \n                                \n                                    Українська мова: 138\n\n                                                                                    Математика: 125\n\n                                                                                    Історія України: 140
3 184  200  200                                \n                                \n                                    Українська мова: 184\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 200
4 184  200  186                                \n                                \n                                    Українська мова: 184\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 186
5 187  200  160                                \n                                \n                                    Українська мова: 187\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 160
6 177  192  180                                \n                          

### 2.3. Трансформація даних у стовпцях

Підготуємо наш набір даних до вигляду, коли у нас будуть стовпці у вигляді

|Rank|Name|StudyType|UkrGrade|MathGrade|HistoryGrade|TotalGrade|IsStudent|
|---|---|---|---|---|---|---|---|
|1|Прізвище І.І.|Бюджет|158|198|184|172.4|TRUE|

`Rank` лишаємо без змін.

#### 2.3.1. Очистка поля `Name`

Очистимо стовпець `Name`. Для початку переглянемо, що насправді може бути у цьому стовпці:

In [10]:
data$Name[1]

[1] "1\n                                                        \n                                120-7964783  В. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1"

Переглянемо скільки символів присутньо у цьому записі

In [11]:
nchar(data$Name)

[1] 213 209 212 208 207 205 208 209 210 209 209 209 208 214 205 209 209 205
 [19] 204 207 213 216 211 208 209 204 208 205 212 208 210 211 207 203 209 207
 [37] 206 208 213 207 210 207 208 209 204 206 212 210 209 203 207 205 204 212
 [55] 210 207 208 209 205 205 211 211 210 211 210 209 209 208 207 208 209 207
 [73] 207 211 208 209 209 203 204 214 208 203 210 208 210 209 210 208 202 209
 [91] 205 210 206 206 208 205 207 204 207 212  50 204 208 210 208 207 211 205
[109] 211 211 205 205 211 209 210 210 212 212 209 210 210 210 207 208 204 210
[127] 203 206 212 212 216 212 215 211 215 211 215 215 213 209 211 212 210 209
[145] 210 210 212 209 211 211 212 208 208 208 211 215 211 210 210 211 212 212
[163] 204 208 212 213 208 209 204 208 209 208 205 214 209 207 211 204 211 208
[181] 209 206 207 204 205 208 210 208 210 210 206

З чого він складається? Цифра на початку, `\n` та `\t`, багато зайвих пробілів та інформація про пріоритет вкінці  `Пр.`. 

Для початку позбудемося `\n` та `\t`.

In [12]:
library(stringr)

In [13]:
name <- data$Name[1]
name <- str_replace_all(name, "\n", "")
name

[1] "1                                                                                        120-7964783  В. В.                                                        \t\t\t\t\t\t                                  Пр. 1"

In [14]:
name <- str_replace_all(name, "\t", "")
name

[1] "1                                                                                        120-7964783  В. В.                                                                                          Пр. 1"

Тепер позбудемося пробілів, наприклад, замінимо 2 пробіли на 1.

In [15]:
str_replace_all(name, "  ", " ")

[1] "1                                            120-7964783 В. В.                                             Пр. 1"

Стало менше проте ми їх не позбулися (((

In [16]:
name <- str_replace_all(name, "  ", " ")
name <- str_replace_all(name, "  ", " ")
name

[1] "1                      120-7964783 В. В.                       Пр. 1"

Майже, але ще не все

In [17]:
name <- str_replace_all(name, "  ", " ")
name <- str_replace_all(name, "  ", " ")
name

[1] "1      120-7964783 В. В.      Пр. 1"

Майже вийшло, проте ми не можемо гарантувати, що усі дані мають таку ж кількість пробілів. Варто написати алгоритм, що буде видаляти усі пробіли за 1 раз.

In [18]:
text_length = 0
text = name

# доки довжина тексту змінюється
while(text_length != nchar(text)) {
    # рахуємо скільки символів у тексті
    text_length <- nchar(text)    
    # замінюємо 2 пробіли на 1
    text <- str_replace_all(text, "  ", " ")   
}
text

[1] "1 120-7964783 В. В. Пр. 1"

Далі загорнемо це все у функцію:

In [19]:
text_cleaner <- function(text) {
    
    text <- str_replace_all(text, "\t", "")
    text <- str_replace_all(text, "\n", "")
    
    text_length = 0

    while(text_length != nchar(text)) {
        text_length <- nchar(text)    
        text <- str_replace_all(text, "  ", " ")   
    }
    text
}

Перевіримо чи функція працює на 1 записі:

In [20]:
text_cleaner(data$Name[3])

[1] "3 Жарчинський М. С. Пр. 1"

Перевіримо чи працює функція у парі з `sapply()` для усього ветора імен

In [21]:
std_names <- sapply(data$Name, text_cleaner)
names(std_names) <- NULL
std_names

[1] "1 120-7964783 В. В. Пр. 1"                         
  [2] "2 Горецька Н. Л. Пр. 1"                            
  [3] "3 Жарчинський М. С. Пр. 1"                         
  [4] "4 Пильгун А. Ю. Пр. 1"                             
  [5] "5 Ошейко В. О. Пр. 2"                              
  [6] "6 Круш І. В. Пр. 4"                                
  [7] "7 Строзюк Р. В. Пр. 1"                             
  [8] "8 Лавренюк А. В. Пр. 1"                            
  [9] "9 Дігалевич І. О. Пр. 1"                           
 [10] "10 Сівачук М. О. Пр. 1"                            
 [11] "11 Свирида А. Ю. Пр. 1"                            
 [12] "12 Мельник Д. А. Пр. 2"                            
 [13] "13 Бондар А. О. Пр. 5"                             
 [14] "14 120-7996285 Б. М. Пр. 1"                        
 [15] "15 Томусяк М. Б. К"                                
 [16] "16 Семенюк В. В. Пр. 1"                            
 [17] "17 Давидюк Р. В. Пр. 1"                            
 [18] "18 Тарасюк Д. О. К"                                
 [19] "19 Довгій Д. М. К"                                 
 [20] "20 Носко Т. В. Пр. 1"                              
 [21] "21 Сокальський О. І. Пр. 1"                        
 [22] "22 Добровольський М. О. Пр. 3"                     
 [23] "23 Літвінчук В. І. Пр. 1"                          
 [24] "24 Шапран Н. В. Пр. 5"                             
 [25] "25 Барковський А. О. К"                            
 [26] "26 Гнатюк І. І. К"                                 
 [27] "27 Бурчак П. О. Пр. 3"                             
 [28] "28 Дзем`юк В. Б. К"                                
 [29] "29 Гладуненко А. Р. Пр. 1"                         
 [30] "30 Хажинський М. С. К"                             
 [31] "31 Мосійчук О. І. Пр. 1"                           
 [32] "32 Ясінський І. О. Пр. 2"                          
 [33] "33 Кошин М. В. Пр. 1"                              
 [34] "34 Дужий А. А. К"                                  
 [35] "35 Семенюк В. С. Пр. 5"                            
 [36] "36 Лук`янчук М. Ю. К"                              
 [37] "37 Іщук М. В. Пр. 1"                               
 [38] "38 Махлай А. М. Пр. 2"                             
 [39] "39 Пухальський В. В. Пр. 2"                        
 [40] "40 Мініч В. В. Пр. 1"                              
 [41] "41 Василець Р. В. Пр. 1"                           
 [42] "42 Ревко А. П. Пр. 3"                              
 [43] "43 Ніколайчук М. А. К"                             
 [44] "44 Поліщук М. С. Пр. 2"                            
 [45] "45 Ковень С. С. К"                                 
 [46] "46 Мостовий В. О. К"                               
 [47] "47 Герасимчук О. С. Пр. 1"                         
 [48] "48 Теренчук Н. І. Пр. 2"                           
 [49] "49 Стерник А. І. Пр. 2"                            
 [50] "50 Мороз В. С. К"                                  
 [51] "51 Максимчук Д. С. К"                              
 [52] "52 Антонюк К. Ю. К"                                
 [53] "53 Бричук А. Т. К"                                 
 [54] "54 Ковінський Д. В. Пр. 1"                         
 [55] "55 Крайчинський М. С. К"                           
 [56] "56 Нестерчук В. Р. К"                              
 [57] "57 Донгаш Є. О. Пр. 4"                             
 [58] "58 Качмарський Д. А. К"                            
 [59] "59 Поліщук М. Б. К"                                
 [60] "60 Рой Д. І. Пр. 1"                                
 [61] "61 Огородник О. С. Пр. 1"                          
 [62] "62 Пахольчак В. С. Пр. 1"                          
 [63] "63 Миронець І. Ю. Пр. 2"                           
 [64] "64 Форманчук Б. І. Пр. 1"                          
 [65] "65 Братасюк Р. В. Пр. 5"                           
 [66] "66 Антонюк А. О. Пр. 2"                            
 [67] "67 Шевцова А. І. Пр. 3"                            
 [68] "68 Єсіпов Д. О. Пр. 5"                    

Як видалити `#Номер` на початку та `Пр. Х` вкінці?

In [22]:
#indexes <- as.data.frame(str_locate(std_names, "Пр."))
#head(indexes, 10)

Варто відмітити, що присутні пропуски у таблиці, а це означає, що `Пр.` не було знайдено у записі. І якщо ми зробимо просто видалення за індексом `Пр.`, то окремі записи будуть знищені. Зверніть увагу на запис `#10`.

In [23]:
#names <- str_sub(std_names, start = 1L, end = indexes$start-2)
#names

Напишемо складніший алгоритм з перевіркою наявності `Пр.` у рядку:

In [24]:
#std_names <- ifelse(is.na(str_sub(std_names, start = 1L, end = indexes$start-2)),
#                    std_names, 
#                    str_sub(std_names, start = 1L, end = indexes$start-2))
#std_names

Далі нам потрібно видалити цифру (порядковий номер) на початку рядка. Алгоритм наступний: розбиваємо на частини рядка через пробіл, а потім усі крім першого "склеюємо" докупи.

In [25]:
x <- unlist(str_split(std_names[1], ' '))
x

[1] "1"           "120-7964783" "В."          "В."          "Пр."        
[6] "1"

In [26]:
x[2:4]

[1] "120-7964783" "В."          "В."

In [27]:
paste(x[2:length(x)], collapse = " ")

[1] "120-7964783 В. В. Пр. 1"

Опишемо це як окрему функцію:

In [28]:
#remove_digit <- function(name) {
#    name <- unlist(str_split(name, ' '))
#    paste(name[2:length(name)], collapse = " ")
#}
#remove_digit(std_names[2])

Щоб виконати цю функцію для усього списку використаємо `lapply()`:

In [29]:
#std_names <- lapply(std_names, remove_digit)
#std_names |> head()

Тепер зберемо докупи усі етапи очистки поля Name в 1 функцію.

In [30]:
#clean_names <- function(names) {
#    student_names <- sapply(names, text_cleaner)
#    names(student_names) <- NULL
#    student_names <- ifelse(is.na(str_sub(student_names, start = 1L, end = indexes$start-2)), 
#                           student_names, str_sub(student_names, start = 1L, end = indexes$start-2))
#    student_names <- lapply(student_names, remove_digit)
#    return(student_names)
#}
get_name_only <- function(name){
    x <- unlist(strsplit(name, " "))[2:4]
    paste(x, collapse = " ")
}
clean_names <- function(names) {
    student_names <- sapply(names, text_cleaner)
    names(student_names) <- NULL
    student_names <- lapply(student_names, get_name_only)
    student_names 
}


In [31]:
clean_names(data$Name) |> head()

[[1]]
[1] "120-7964783 В. В."

[[2]]
[1] "Горецька Н. Л."

[[3]]
[1] "Жарчинський М. С."

[[4]]
[1] "Пильгун А. Ю."

[[5]]
[1] "Ошейко В. О."

[[6]]
[1] "Круш І. В."

Замінимо дані у дата-фреймі:

In [32]:
data <- data |>
    mutate(Name = clean_names(Name))
data |> head()

Rank Name              TotalGrade StudyType    
1 1    120-7964783 В. В. 176.100    До наказу (б)
2 2    Горецька Н. Л.    131.900    До наказу (б)
3 3    Жарчинський М. С. 195.200    До наказу (б)
4 4    Пильгун А. Ю.     192.400    До наказу (б)
5 5    Ошейко В. О.      188.100    До наказу (б)
6 6    Круш І. В.        185.100    До наказу (б)
  Grades                                                                                                                                                                                                                                                                                                                                                      
1 187  176  160                                \n                                \n                                    Українська мова: 187\n\n                                                                                    Математика: 176\n\n                                                                                    Історія України: 160
2 138  125  140                                \n                                \n                                    Українська мова: 138\n\n                                                                                    Математика: 125\n\n                                                                                    Історія України: 140
3 184  200  200                                \n                                \n                                    Українська мова: 184\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 200
4 184  200  186                                \n                                \n                                    Українська мова: 184\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 186
5 187  200  160                                \n                                \n                                    Українська мова: 187\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 160
6 177  192  180                                \n                                \n                                    Українська мова: 177\n\n                                                                                    Математика: 192\n\n                                                                                    Історія України: 180
  IsStudent
1 +        
2 +        
3 +        
4 +        
5 +        
6 +

#### 2.3.2. Робота з полем `StudyType`

In [33]:
unique(data$StudyType)

[1] "До наказу (б)"                                     
[2] "До наказу (к)"                                     
[3] "Допущено"                                          
[4] "(adsbygoogle = window.adsbygoogle || []).push({});"
[5] ""                                                  
[6] "Скасовано (втрата пріор.)"

In [34]:
as.vector(data$StudyType)

[1] "До наказу (б)"                                     
  [2] "До наказу (б)"                                     
  [3] "До наказу (б)"                                     
  [4] "До наказу (б)"                                     
  [5] "До наказу (б)"                                     
  [6] "До наказу (б)"                                     
  [7] "До наказу (б)"                                     
  [8] "До наказу (б)"                                     
  [9] "До наказу (б)"                                     
 [10] "До наказу (б)"                                     
 [11] "До наказу (б)"                                     
 [12] "До наказу (б)"                                     
 [13] "До наказу (б)"                                     
 [14] "До наказу (к)"                                     
 [15] "До наказу (к)"                                     
 [16] "До наказу (к)"                                     
 [17] "До наказу (к)"                                     
 [18] "До наказу (к)"                                     
 [19] "До наказу (к)"                                     
 [20] "До наказу (к)"                                     
 [21] "До наказу (к)"                                     
 [22] "До наказу (к)"                                     
 [23] "До наказу (к)"                                     
 [24] "До наказу (к)"                                     
 [25] "До наказу (к)"                                     
 [26] "До наказу (к)"                                     
 [27] "До наказу (к)"                                     
 [28] "До наказу (к)"                                     
 [29] "До наказу (к)"                                     
 [30] "До наказу (к)"                                     
 [31] "До наказу (к)"                                     
 [32] "До наказу (к)"                                     
 [33] "До наказу (к)"                                     
 [34] "До наказу (к)"                                     
 [35] "До наказу (к)"                                     
 [36] "До наказу (к)"                                     
 [37] "До наказу (к)"                                     
 [38] "До наказу (к)"                                     
 [39] "До наказу (к)"                                     
 [40] "До наказу (к)"                                     
 [41] "До наказу (к)"                                     
 [42] "До наказу (к)"                                     
 [43] "До наказу (к)"                                     
 [44] "До наказу (к)"                                     
 [45] "До наказу (к)"                                     
 [46] "До наказу (к)"                                     
 [47] "До наказу (к)"                                     
 [48] "До наказу (к)"                                     
 [49] "До наказу (к)"                                     
 [50] "До наказу (к)"                                     
 [51] "До наказу (к)"                                     
 [52] "До наказу (к)"                                     
 [53] "До наказу (к)"                                     
 [54] "До наказу (к)"                                     
 [55] "До наказу (к)"                                     
 [56] "До наказу (к)"                                     
 [57] "До наказу (к)"                                     
 [58] "До наказу (к)"                                     
 [59] "До наказу (к)"                                     
 [60] "До наказу (к)"                                     
 [61] "До наказу (к)"                                     
 [62] "До наказу (к)"                                     
 [63] "До наказу (к)"                                     
 [64] "До наказу (к)"                                     
 [65] "Допущено"                                          
 [66] "Допущено"                                          
 [67] "Допущено"                                          
 [68] "Допущено"                                 

In [35]:
data <- data |>
    mutate(StudyType = ifelse(StudyType == "До наказу (б)", "Бюджет",
                              ifelse(StudyType == "До наказу (к)", "Контракт",
                                     "Не навчається"))       
)
data |> head()

Rank Name              TotalGrade StudyType
1 1    120-7964783 В. В. 176.100    Бюджет   
2 2    Горецька Н. Л.    131.900    Бюджет   
3 3    Жарчинський М. С. 195.200    Бюджет   
4 4    Пильгун А. Ю.     192.400    Бюджет   
5 5    Ошейко В. О.      188.100    Бюджет   
6 6    Круш І. В.        185.100    Бюджет   
  Grades                                                                                                                                                                                                                                                                                                                                                      
1 187  176  160                                \n                                \n                                    Українська мова: 187\n\n                                                                                    Математика: 176\n\n                                                                                    Історія України: 160
2 138  125  140                                \n                                \n                                    Українська мова: 138\n\n                                                                                    Математика: 125\n\n                                                                                    Історія України: 140
3 184  200  200                                \n                                \n                                    Українська мова: 184\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 200
4 184  200  186                                \n                                \n                                    Українська мова: 184\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 186
5 187  200  160                                \n                                \n                                    Українська мова: 187\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 160
6 177  192  180                                \n                                \n                                    Українська мова: 177\n\n                                                                                    Математика: 192\n\n                                                                                    Історія України: 180
  IsStudent
1 +        
2 +        
3 +        
4 +        
5 +        
6 +

#### 2.3.3. Очистка інформації про оцінки (`Grades`)

Як бачимо, поле `Grades` містить інформацію про усі оцінки + нави предметів. Наша задача - розділити ці дані на 3 окремі стовпці з інформацією лише про оцінки, а назви предметів варто закодувати у назвах стовпців.

In [36]:
data |> select(Grades) |> top_n(3)

Selecting by Grades


Grades                                                                                                                                                                                                                                                                                                                                                      
1 200  200  192                                \n                                \n                                    Українська мова: 200\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 192
2 194  200  200                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 200
3 194  200  200                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 200

In [37]:
data$Grades[1]

[1] "187  176  160                                \n                                \n                                    Українська мова: 187\n\n                                                                                    Математика: 176\n\n                                                                                    Історія України: 160"

Оцінимо як виглядає 1 запис. Дуже схожі проблеми з полем Name. Спробуємо примінити уже готову функцію text_cleaner до цього поля.

In [38]:
grades <- text_cleaner(data$Grades[1])
grades

[1] "187 176 160 Українська мова: 187 Математика: 176 Історія України: 160"

Вийшло. Наступний етап розбиття інформації на частини. Нам потрібні лише 3 перші записи у цьому полі.

In [39]:
grades <- unlist(strsplit(grades, " "))[1:3]
grades

[1] "187" "176" "160"

In [40]:
sapply(grades, text_cleaner) |> head()

187   176   160 
"187" "176" "160"

In [41]:
get_3_grades <- function(grades){
    unlist(strsplit(grades, " "))[1:3]
}

get_grades <- function(grades) {
    grades <- sapply(grades, text_cleaner)
    names(grades) <- NULL
    grades <- lapply(grades, get_3_grades)
    grades <- lapply(grades, as.numeric)
    grades
}  

get_grades(data$Grades) |> head()

Warning message in lapply(grades, as.numeric):
"NAs introduced by coercion"


[[1]]
[1] 187 176 160

[[2]]
[1] 138 125 140

[[3]]
[1] 184 200 200

[[4]]
[1] 184 200 186

[[5]]
[1] 187 200 160

[[6]]
[1] 177 192 180

In [42]:
grades <- get_grades(data$Grades)
grades |> head()

Warning message in lapply(grades, as.numeric):
"NAs introduced by coercion"


[[1]]
[1] 187 176 160

[[2]]
[1] 138 125 140

[[3]]
[1] 184 200 200

[[4]]
[1] 184 200 186

[[5]]
[1] 187 200 160

[[6]]
[1] 177 192 180

In [43]:
data <- data |>
    mutate(UkrGrades = unlist(lapply(grades, `[[`, 1)),
           MathGrades = unlist(lapply(grades, `[[`, 2)),
           HistoryGrades = unlist(lapply(grades, `[[`, 3)),
           Grades = NULL)
data |> head()

Rank Name              TotalGrade StudyType IsStudent UkrGrades MathGrades
1 1    120-7964783 В. В. 176.100    Бюджет    +         187       176       
2 2    Горецька Н. Л.    131.900    Бюджет    +         138       125       
3 3    Жарчинський М. С. 195.200    Бюджет    +         184       200       
4 4    Пильгун А. Ю.     192.400    Бюджет    +         184       200       
5 5    Ошейко В. О.      188.100    Бюджет    +         187       200       
6 6    Круш І. В.        185.100    Бюджет    +         177       192       
  HistoryGrades
1 160          
2 140          
3 200          
4 186          
5 160          
6 180

#### 2.3.4. Трансформація поля `IsStudent`

Замінимо це поле на логічне та змінимо порядок стовпців:

In [44]:
data <- data |>
    mutate(IsStudent = ifelse(IsStudent == "+", T, F)) |>
    select(1:2, 6:8, 3:5)

In [45]:
data |> head(15)

Rank Name              UkrGrades MathGrades HistoryGrades TotalGrade
1  1    120-7964783 В. В. 187       176        160           176.100   
2  2    Горецька Н. Л.    138       125        140           131.900   
3  3    Жарчинський М. С. 184       200        200           195.200   
4  4    Пильгун А. Ю.     184       200        186           192.400   
5  5    Ошейко В. О.      187       200        160           188.100   
6  6    Круш І. В.        177       192        180           185.100   
7  7    Строзюк Р. В.     174       192        180           184.200   
8  8    Лавренюк А. В.    194       185        162           183.100   
9  9    Дігалевич І. О.   172       192        176           182.800   
10 10   Сівачук М. О.     187       192        151           182.300   
11 11   Свирида А. Ю.     169       185        180           179.200   
12 12   Мельник Д. А.     190       172        173           177.600   
13 13   Бондар А. О.      190       168        176           176.200   
14 14   120-7996285 Б. М. 137       138        140           138.100   
15 15   Томусяк М. Б.     187       185        164           181.400   
   StudyType IsStudent
1  Бюджет    TRUE     
2  Бюджет    TRUE     
3  Бюджет    TRUE     
4  Бюджет    TRUE     
5  Бюджет    TRUE     
6  Бюджет    TRUE     
7  Бюджет    TRUE     
8  Бюджет    TRUE     
9  Бюджет    TRUE     
10 Бюджет    TRUE     
11 Бюджет    TRUE     
12 Бюджет    TRUE     
13 Бюджет    TRUE     
14 Контракт  TRUE     
15 Контракт  TRUE

In [46]:
mean(data$UkrGrades, na.rm = T)

[1] 160.6053

In [47]:
data |>
    group_by(IsStudent) |>
    summarise(UkrMean = mean(UkrGrades, na.rm = T),
             MathMean = mean(MathGrades, na.rm = T),
             HistoryMean = mean(HistoryGrades, na.rm = T))

IsStudent UkrMean  MathMean HistoryMean
1 FALSE     162.7381 163.0635 159.1746   
2  TRUE     156.4062 157.6406 155.1562

In [48]:
abit_parser <- function(url) {
    
    page <- read_html(url)    
    tables <- html_nodes(page, "table")    
    
    data <- as.data.frame(html_table(tables[1]), fill = T) |> select(-c(3,7))
    
    colnames(data) <- c("Rank", "Name", "TotalGrade", "StudyType", "Grades", "IsStudent")
    
    grades <- get_grades(data$Grades)
    
    data <- data |> 
        mutate(Name = clean_names(Name),
               StudyType = ifelse(StudyType == "До наказу (б)", "Бюджет",
                                  ifelse(StudyType == "До наказу (к)", "Контракт",
                                         "Не навчається")),
               UkrGrades = unlist(lapply(grades, `[[`, 1)),
               MathGrades = unlist(lapply(grades, `[[`, 2)),
               HistoryGrades = unlist(lapply(grades, `[[`, 3)),
               Grades = NULL, 
               IsStudent = ifelse(IsStudent == "+", T, F)) |>
    select(1:2, 6:8, 3:5)
    
    data
}

In [49]:
data_new <- abit_parser("https://abit-poisk.org.ua/rate2022/direction/984185")

Warning message in lapply(grades, as.numeric):
"NAs introduced by coercion"


In [51]:
data_new 

Rank Name                 UkrGrades MathGrades HistoryGrades TotalGrade
1   1    120-7964783 В. В.    187       176        160           176.100   
2   2    Горецька Н. Л.       138       125        140           131.900   
3   3    Жарчинський М. С.    184       200        200           195.200   
4   4    Пильгун А. Ю.        184       200        186           192.400   
5   5    Ошейко В. О.         187       200        160           188.100   
6   6    Круш І. В.           177       192        180           185.100   
7   7    Строзюк Р. В.        174       192        180           184.200   
8   8    Лавренюк А. В.       194       185        162           183.100   
9   9    Дігалевич І. О.      172       192        176           182.800   
10  10   Сівачук М. О.        187       192        151           182.300   
11  11   Свирида А. Ю.        169       185        180           179.200   
12  12   Мельник Д. А.        190       172        173           177.600   
13  13   Бондар А. О.         190       168        176           176.200   
14  14   120-7996285 Б. М.    137       138        140           138.100   
15  15   Томусяк М. Б.        187       185        164           181.400   
16  16   Семенюк В. В.        172       176        176           174.800   
17  17   Давидюк Р. В.        182       168        167           172.000   
18  18   Тарасюк Д. О.        145       192        158           171.100   
19  19   Довгій Д. М.         154       185        158           170.300   
20  20   Носко Т. В.          157       176        160           167.100   
21  21   Сокальський О. І.    172       159        176           166.300   
22  22   Добровольський М. О. 160       168        156           163.200   
23  23   Літвінчук В. І.      151       176        146           162.500   
24  24   Шапран Н. В.         154       168        154           161.000   
25  25   Барковський А. О.    177       154        153           160.700   
26  26   Гнатюк І. І.         177       149        164           160.400   
27  27   Бурчак П. О.         145       165        164           158.800   
28  28   Дзем`юк В. Б.        149       159        170           158.200   
29  29   Гладуненко А. Р.     160       156        156           157.200   
30  30   Хажинський М. С.     169       149        156           156.400   
... ...  ...                  ...       ...        ...           ...       
162 161  Дімітрова А. О.      190       168        170           175.000   
163 162  Журба А. С.          180       176        164           174.800   
164 163  Білая Є. С.          184       172        167           174.600   
165 164  Старостюк В. С.      187       176        151           174.300   
166 165  Степовенко Ю. О.     172       165        200           174.100   
167 166  Сисюк В. В.          190       172        151           173.200   
168 167  Шелест О. О.         169       172        180           172.700   
169 168  Луцук В. Ю.          157       192        147           172.500   
170 169  Луцук В. Ю.          157       192        147           172.500   
171 170  Панько П. Т.         174       159        180           167.700   
172 171  Білий О. В.          154       176        167           167.600   
173 172  Шергей М. В.         169       168        160           166.700   
174 173  Грабовський П. І.    180       168        140           166.000   
175 174  Ткачук Д. В.         160       172        160           166.000   
176 175  Політько С. В.       180       159        160           165.500   
177 176  Будяківський М. Ю.   160       176        146           165.200   
178 177  Федас Р. В.          160       172        154           164.800   
179 178  Коханець В. І.       163       159        176           163.600   
180 179  Рудюк О. О.          169       159        162           162.600   
181 180  Криворучко А. М.     180       152        160           162.000   
182 181  Романюк О. В.        169       162        1